contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,155 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.10;
+import "../libraries/Address.sol";
+
+// types
+import "../types/BaseAllocator.sol";
+
+// interfaces
+import "../interfaces/IMasterChef.sol";
+import "../interfaces/ISushiBar.sol";
+
+/**
+ *  Contract deploys liquidity from treasury into the Onsen program,
+ *  earning $SUSHI that can be staked and/or deposited into the treasury.
+ */
+contract OnsenAllocatorV2 is BaseAllocator {
+    using SafeERC20 for IERC20;
+
+    /* ========== STATE VARIABLES ========== */
+
+    address immutable sushi; // $SUSHI token
+    address immutable xSushi; // $xSUSHI token
+
+    address immutable masterChef; // Onsen contract
+
+    // address immutable treasury;
+    address public constant treasury = 0x9A315BdF513367C0377FB36545857d12e85813Ef; // Olympus Treasury
+
+    /* ========== CONSTRUCTOR ========== */
+
+    constructor(
+        address _chef,
+        address _sushi,
+        address _xSushi,
+        AllocatorInitData memory data
+    ) BaseAllocator(data) {
+        require(_chef != address(0));
+        masterChef = _chef;
+        require(_sushi != address(0));
+        sushi = _sushi;
+        require(_xSushi != address(0));
+        xSushi = _xSushi;
+
+        // approve for safety, yes toke is being instantly sent to treasury and that is fine
+        // but to be absolutely safe this one approve won't hurt
+        IERC20(sushi).approve(address(extender), type(uint256).max);
+        IERC20(xSushi).approve(address(extender), type(uint256).max);
+    }
+
+    /**
+     1- Deposit LP token into onsen, if deposit is succesfull we should get in return sushi tokens
+     2- Stake the Sushi rewards tokens returned from deposit LP tokens and as rewards form onsen LP stake, we should get in return xSushi tokens
+     3- Calculate gains/loss
+     */
+    function _update(uint256 id) internal override returns (uint128 gain, uint128 loss) {
+        // reads
+        uint256 index = tokenIds[id];
+        IERC20 LPtoken = _tokens[index];
+        uint256 balance = LPtoken.balanceOf(address(this));
+
+        // interactions
+        //Deposit LP token into onsen, if deposit succesfull this address should have in return sushi tokens
+        if (balance > 0) {
+            IERC20(LPtoken).approve(masterChef, balance);
+            IMasterChef(masterChef).deposit(id, balance); // deposit into Onsen
+        }
+
+        // Stake the sushi tokens
+        if (IERC20(sushi).balanceOf(address(this)) > 0) {
+            enterSushiBar(true); // manage rewards
+        }
+
+        //Calculate gains/loss
+        // Retrieve current balance for pool and address
+        UserInfo memory currentUserInfo = IMasterChef(masterChef).userInfo(id, address(this));
+
+        uint256 last = extender.getAllocatorAllocated(id) + extender.getAllocatorPerformance(id).gain;
+
+        if (currentUserInfo.amount >= last) {
+            gain = uint128(currentUserInfo.amount - last);
```

Comment:
so since theres no compounding and LP's gain and lose value from being worth more or less due to fees and IL. Realistically we won't ever see any gains or losses right since the amount of LP tokens will always stay the same.

Reply:
@Paul4912  Yes that's correct there shouldn't be any difference for the LP amount, but not sure if we should still do the comparison? or should we remove it and always return 0 for gain and loss?

Reply:
You need to know if you the LP tokens are constant balance and increasing value or vice versa, and if the former note down the increasing balance, since the LP tokens are being withdrawn from the treasury and this needs to be registered sir. Still, always note down the LP balance please, since there might be a contract failure and you need to report that so the allocator can panic and potentially do something in case where all tokens have not been lost.

Reply:
Gain/Loss balance is still been returned based on the balance for the LP tokens in the Onsen pool.

LP tokens balance is constant, since rewards are awarded with sushi tokens